sql/wsrep_schema.cc

```diff
@@ -1535,3 +1555,61 @@ int Wsrep_schema::recover_sr_transactions(THD *orig_thd)
 out:
   DBUG_RETURN(ret);
 }
+
+int Wsrep_schema::store_gtid_event(THD* thd,
+                                   const Gtid_log_event *gtid)
+{
+  DBUG_ENTER("Wsrep_schema::store_gtid_event");
+  int error=0;
+  void *hton= NULL;
+  const bool in_transaction= (gtid->flags2 & Gtid_log_event::FL_TRANSACTIONAL);
+  const bool in_ddl= (gtid->flags2 & Gtid_log_event::FL_DDL);
+
+  DBUG_PRINT("info", ("thd: %p, in_transaction: %d, in_ddl: %d "
+                      "in_active_multi_stmt_transaction: %d",
+                      thd, in_transaction, in_ddl,
+                      thd->in_active_multi_stmt_transaction()));
+
+  Wsrep_schema_impl::wsrep_ignore_table  ignore_table(thd);
+  Wsrep_schema_impl::binlog_off binlog_off(thd);
+  Wsrep_schema_impl::sql_safe_updates sql_safe_updates(thd);
+
+  rpl_group_info *rgi= thd->wsrep_rgi;
+  const uint64 sub_id= rpl_global_gtid_slave_state->next_sub_id(gtid->domain_id);
+  rpl_gtid current_gtid;
+  current_gtid.domain_id= gtid->domain_id;
+  current_gtid.server_id= gtid->server_id;
+  current_gtid.seq_no= gtid->seq_no;
+  rgi->gtid_pending= false;
+
+  DBUG_ASSERT(!in_transaction || thd->in_active_multi_stmt_transaction());
+
+  if ((error= rpl_global_gtid_slave_state->record_gtid(thd, &current_gtid,
+                                                      sub_id,
+                                                      in_transaction, false, &hton)))
+    goto out;
+
+  rpl_global_gtid_slave_state->update_state_hash(sub_id, &current_gtid, hton, rgi);
+
+  if (in_ddl)
+  {
+    // Commit transaction if this GTID is part of DDL-clause because
+    // DDL causes implicit commit assuming there is no multi statement
+    // transaction ongoing.
+    if((error= trans_commit_stmt(thd)))
+      goto out;
+
+    (void)trans_commit(thd);
```

Comment:
I'm not really familiar with the Galera code - when in the transaction life-cycle is `store_gtid_event()`/`wsrep_apply_events()` called? It looks to me like this is before any of the transaction events have actually been applied. If so, and then we update & commit `mysql.gtid_slave_pos` here before actually applying the DDL, then if something happens while the DDL is running and we restart, `gtid_slave_pos` would be ahead of the actual database state.

Reply:
In case of DDL, it is problematic. If GTID event is executed before DDL, it will be committed anyway as DDL does implicit commit before start. IF GTID event is executed after DDL it is also not atomic with DDL.